lib/node_modules/@stdlib/string/base/last-code-point/lib/main.js

```diff
@@ -0,0 +1,99 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var RE_UTF16_SURROGATE_PAIR = require( '@stdlib/regexp/utf16-surrogate-pair' ).REGEXP;
+
+
+// VARIABLES //
+
+var RE_UTF16_LOW_SURROGATE = /[\uDC00-\uDFFF]/; // TODO: replace with stdlib pkg
+var RE_UTF16_HIGH_SURROGATE = /[\uD800-\uDBFF]/; // TODO: replace with stdlib pkg
+
+
+// MAIN //
+
+/**
+* Returns the last `n` Unicode code points of a string.
+*
+* @param {string} str - input string
+* @param {NonNegativeInteger} n - number of Unicode code points to return
+* @returns {string} output string
+*
+* @example
+* var out = lastCodePoint( 'Hello World', 1 );
+* // returns 'd'
+*
+* @example
+* var out = lastCodePoint( 'ASCII', 2 );
+* // returns 'II'
+*
+* @example
+* var out = lastCodePoint( 'JavaScript', 6 );
+* // returns 'Script'
+*
+*/
+function lastCodePoint( str, n ) {
+	var len;
+	var out;
+	var ch1;
+	var ch2;
+	var cnt;
+	var i;
+	len = str.length;
+	out = '';
+	cnt = 0;
+	if ( str === '' || n === 0 ) {
+		return '';
+	}
+	if ( n === 1 ) {
+		str = str.substring( len-2, len );
+		if ( RE_UTF16_SURROGATE_PAIR.test( str ) ) {
+			return str;
+		}
+		return str[1];
+	}
+	for ( i = len-1; i >= 0; i-- ) {
+		ch1 = str[ i ];
+		out = ch1 + out;
+		cnt += 1;
+
+		if ( RE_UTF16_LOW_SURROGATE.test( ch1 ) ) {
+			if ( i === 0 ) {
+				break;
+			}
+			ch2 = str[ i+1 ];
```

Comment:
@adityacodes30 Please double-check the implementation; shouldn't line 83 be the below?

```suggestion
			ch2 = str[ i-1 ];
```

It may be good to add some more test cases here.

Reply:
yes, it should be - strange that the test cases pass, maybe its because we havent tested for strings/chars that dont exhibit full surrogate pairs

Reply:
i've however added as many edge cases as i can think of, let me know if you want me to think in a particular direction

Reply:
@adityacodes30 Try adding test cases to see whether surrogate pairs are handled correctly, e.g.

```
    t.equal(last('😄', 1), '😄', 'should handle surrogate pairs correctly');
    t.equal(last('😊👍', 1), '👍', 'should handle multiple surrogate pairs correctly');
    t.equal(last('abc😄def', 4), '😄def', 'should handle surrogate pairs mixed with regular characters');
    t.equal(last('😄😄😄', 2), '😄😄', 'should handle multiple consecutive surrogate pairs');
```

I think there may still be a bug in there.

Reply:
i did add the cases with composite strings , let me test with these as well